service/validation.go

```diff
@@ -1,97 +1,117 @@
-		if warning != nil {
-			warnings = append(warnings, warning)
-		}
+func (p *parameterValidator) Validate(params []*Parameter, data interface{}) ([]*ParameterWarning, error) {
+	v := vtree.Analyze(data)
+	if v.Type != vtree.Object {
+		return nil, errors.New("not an object")
 	}
-
-	return warnings
-}
-
-// parameterValidator provides functionalities to check data against its parameter schema.
-type parameterValidator struct {
-	parameter *Parameter
+	p.validateParameters(params, v)
+	return p.warnings, nil
 }
 
-func newParameterValidator(parameter *Parameter) *parameterValidator {
-	return &parameterValidator{parameter}
-}
-
-// Validate validates value by comparing to its parameter schema.
-func (v *parameterValidator) Validate(value interface{}) *ParameterWarning {
-	if value == nil {
-		if v.parameter.Optional {
-			return nil
+func (p *parameterValidator) validateParameters(params []*Parameter, v vtree.Value) {
+	for _, param := range params {
+		v, ok := v.GetByKey(param.Key, false)
+		if !ok {
+			p.checkOptional(param)
+			continue
 		}
-		return v.newParameterWarning("required")
+		p.validateParameter(param, v)
 	}
-
-	return v.validateType(value)
 }
 
-// validateType checks if value comforts its expected type.
-func (v *parameterValidator) validateType(value interface{}) *ParameterWarning {
-	switch v.parameter.Type {
-	case "String":
-		if _, ok := value.(string); !ok {
-			return v.newParameterWarning("not a string")
-		}
-
-	case "Number":
-		_, okFloat64 := value.(float64)
-		_, okFloat32 := value.(float32)
-		_, okInt := value.(int)
-		if !okInt && !okFloat64 && !okFloat32 {
-			return v.newParameterWarning("not a number")
+func (p *parameterValidator) validateParameter(param *Parameter, v vtree.Value) {
+	if param.Repeated {
+		if v.Type == vtree.Array {
+			for _, vv := range v.Values {
+				p.validateParameterType(param, vv)
+			}
+		} else {
+			p.addWarning(param, "not an array")
 		}
+	} else {
+		p.validateParameterType(param, v)
+	}
+}
 
-	case "Boolean":
-		if _, ok := value.(bool); !ok {
-			return v.newParameterWarning("not a boolean")
-		}
+// validateParameterType checks parameter's type.
+func (p *parameterValidator) validateParameterType(param *Parameter, v vtree.Value) {
+	if v.Type == vtree.Nil {
+		p.checkOptional(param)
+		return
+	}
 
-	case "Object":
-		_, okObj := value.(map[string]interface{})
-		_, okArr := value.([]interface{})
-		if !okObj && !okArr {
-			return v.newParameterWarning("not an object or array")
+	switch {
+	// interpret basic types.
+	case param.Type != "":
+		switch param.Type {
+		case "String":
+			if v.Type != vtree.String {
+				p.addWarning(param, "not a string")
+			}
+		case "Number":
+			if v.Type != vtree.Number {
+				p.addWarning(param, "not a number")
+			}
+		case "Boolean":
+			if v.Type != vtree.Bool {
+				p.addWarning(param, "not a boolean")
+			}
+		// TODO(ilgooz) Deprecate Object in future.
+		case "Any", "Object":
```

Comment:
I suggest to make a breaking change and directly force to remove Object from service definition.